src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/SamlSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,243 @@
+        { }
+
+        public SamlSecurityTokenAuthenticator(IList<SecurityTokenAuthenticator> supportingAuthenticators, TimeSpan maxClockSkew)
+        {
+
+            if (supportingAuthenticators == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(supportingAuthenticators));
+
+            this._supportingAuthenticators = new List<SecurityTokenAuthenticator>(supportingAuthenticators.Count);
+            for (int i = 0; i < supportingAuthenticators.Count; ++i)
+            {
+                this._supportingAuthenticators.Add(supportingAuthenticators[i]);
+            }
+
+            _maxClockSkew = maxClockSkew;
+            _audienceUriMode = AudienceUriMode.Always;
+            _allowedAudienceUris = new Collection<string>();
+        }
+
+        public AudienceUriMode AudienceUriMode
+        {
+            get { return _audienceUriMode; }
+            set
+            {
+                AudienceUriModeValidationHelper.Validate(_audienceUriMode);
+                _audienceUriMode = value;
+            }
+        }
+
+        public IList<string> AllowedAudienceUris
+        {
+            get { return _allowedAudienceUris; }
+        }
+
+        protected override bool CanValidateTokenCore(SecurityToken token)
+        {
+            return token is SamlSecurityToken;
+        }
+
+        protected override ValueTask<ReadOnlyCollection<IAuthorizationPolicy>> ValidateTokenCoreAsync(Tokens.SecurityToken token)
+        {
+            if (token == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(token));
+
+            SamlSecurityToken samlToken = token as SamlSecurityToken;
+
+            if (samlToken == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenAuthenticatorCanOnlyProcessSamlTokens, token.GetType().ToString())));
+
+            if (samlToken.Assertion.Signature == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenMissingSignature)));
+
+            if (!IsCurrentlyTimeEffective(samlToken))
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLTokenTimeInvalid, DateTime.UtcNow.ToUniversalTime(), samlToken.ValidFrom.ToString(), samlToken.ValidTo.ToString())));
+
+          //  if (samlToken.Assertion.SigningToken == null)
+          //      throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlSigningTokenMissing)));
+
+            // Build the Issuer ClaimSet for this Saml token.
+            ClaimSet issuer = null;
+            bool canBeValidated = false;
+            for (int i = 0; i < _supportingAuthenticators.Count; ++i)
+            {
+                canBeValidated = _supportingAuthenticators[i].CanValidateToken(samlToken);
+                if (canBeValidated)
```

Comment:
This is having the supporting authenticator check the Xml element of this token and we're not validating the signing token. So for example, if the Saml token has been signed by an X509 certificate, the SigningToken will be an X509SecurityToken.

Reply:
discussed and incorporated